src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java

```diff
@@ -0,0 +1,229 @@
+import com.google.googleidentity.oauth2.request.OAuth2Request;
+import com.google.googleidentity.oauth2.request.RequestHandler;
+import com.google.googleidentity.oauth2.token.OAuth2AccessToken;
+import com.google.googleidentity.oauth2.token.OAuth2TokenService;
+import com.google.googleidentity.oauth2.util.OAuth2ParameterNames;
+import com.google.googleidentity.user.UserDetails;
+import com.google.googleidentity.user.UserDetailsService;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Jws;
+import io.jsonwebtoken.JwtException;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SigningKeyResolverAdapter;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.servlet.http.HttpServletResponse;
+import net.minidev.json.JSONObject;
+import org.apache.http.HttpStatus;
+
+/** Processor for dealing JWT Assertion */
+@Singleton
+public class JwtAssertionRequestHandler implements RequestHandler {
+
+  private static final String keyUrl = "https://www.googleapis.com/oauth2/v3/certs";
+  private static final String aud =
+      "475640046628-i42g5qfbcp58e3nijqiedomvhe7hb3sn.apps.googleusercontent.com";
+  private final OAuth2TokenService oauth2TokenService;
+  private final UserDetailsService userDetailsService;
+
+  private final ClientDetailsService clientDetailsService;
+
+  private final Logger log = Logger.getLogger("JwtAssertionTokenProcessor");
+
+  @Inject
+  public JwtAssertionRequestHandler(
+      OAuth2TokenService oauth2TokenService,
+      UserDetailsService userDetailsService,
+      ClientDetailsService clientDetailsService) {
+    this.oauth2TokenService = oauth2TokenService;
+    this.userDetailsService = userDetailsService;
+    this.clientDetailsService = clientDetailsService;
+  }
+
+  @Override
+  public void handle(HttpServletResponse response, OAuth2Request oauth2Request)
+      throws IOException, OAuth2Exception {
+    List<String> info =
+        verifyAndGetInfoFromJwt(
+            oauth2Request.getRequestBody().getAssertion(), new JwtSigningKeyResolver(keyUrl));
+    String email = info.get(0);
+    String googleAccountId = info.get(1);
+    switch (oauth2Request.getRequestBody().getIntent()) {
+      case CHECK:
+        handleCheckAssertion(response, email, googleAccountId);
+        break;
+      case GET:
+        handleGetAssertion(
+            response, email, googleAccountId, oauth2Request.getRequestBody().getScopesList());
+        break;
+      case CREATE:
+        handleCreateAssertion(
+            response, email, googleAccountId, oauth2Request.getRequestBody().getScopesList());
+        break;
+      default:
+        throw new IllegalStateException();
+    }
+  }
+
+  public List<String> verifyAndGetInfoFromJwt(String assertion, SigningKeyResolverAdapter keys)
```

Comment:
Can this be private?

Reply:
I set it to public for tests.

Reply:
Then you can make it package private and annotate as @VisibleForTesting